### ISSUE TITLE
`karmadactl init` - full verification and repair option

### ISSUE THREAD

**Issue by tillig** [karmada-io/karmada#3677]
**What would you like to be added**:

When I run `karmadactl init` on a cluster, if the cluster is already initialized it'd be nice to have the `init` verify that things are wired up correctly and fix things that aren't.

**Why is this needed**:

Right now if the `init` fails for some reason, you can only `deinit` and try again.

I think my corporate firewall is blocking some ports, so my current `init` is failing:

```text
I0616 14:04:17.151424   67476 deploy.go:177] kubeconfig file: , kubernetes: https://XXXXXXX.hcp.westus2.azmk8s.io:443
W0616 14:04:18.515494   67476 node.go:36] The kubernetes cluster does not have a Master role.
I0616 14:04:18.515524   67476 node.go:44] Randomly select 3 Node IPs in the kubernetes cluster.
I0616 14:04:18.667315   67476 deploy.go:197] karmada apiserver ip: [10.1.0.4 10.1.0.33]
I0616 14:04:19.431719   67476 cert.go:229] Generate ca certificate success.
I0616 14:04:19.892572   67476 cert.go:229] Generate karmada certificate success.
I0616 14:04:20.139472   67476 cert.go:229] Generate apiserver certificate success.
I0616 14:04:20.350338   67476 cert.go:229] Generate front-proxy-ca certificate success.
I0616 14:04:20.395515   67476 cert.go:229] Generate front-proxy-client certificate success.
I0616 14:04:20.451684   67476 cert.go:229] Generate etcd-ca certificate success.
I0616 14:04:20.607064   67476 cert.go:229] Generate etcd-server certificate success.
I0616 14:04:20.692963   67476 cert.go:229] Generate etcd-client certificate success.
I0616 14:04:20.694261   67476 deploy.go:291] download crds file:https://github.com/karmada-io/karmada/releases/download/v1.6.0/crds.tar.gz
Downloading...[ 100.00% ]
Download complete.
I0616 14:04:21.437378   67476 deploy.go:531] Create karmada kubeconfig success.
I0616 14:04:21.865539   67476 idempotency.go:251] Namespace karmada-system has been created or updated.
I0616 14:04:22.765526   67476 idempotency.go:275] Service karmada-system/etcd has been created or updated.
I0616 14:04:22.765566   67476 deploy.go:359] Create etcd StatefulSets
I0616 14:04:26.215734   67476 deploy.go:367] Create karmada ApiServer Deployment
I0616 14:04:26.365473   67476 idempotency.go:275] Service karmada-system/karmada-apiserver has been created or updated.
I0616 14:04:57.813269   67476 deploy.go:382] Create karmada aggregated apiserver Deployment
I0616 14:04:57.962586   67476 idempotency.go:275] Service karmada-system/karmada-aggregated-apiserver has been created or updated.
F0616 14:05:01.479656   67476 deploy.go:57] unable to create Namespace: Post "https://10.1.0.4:32443/api/v1/namespaces": dial tcp 10.1.0.4:32443: connect: no route to host
```

I'm not totally sure what's up there, but I can't rerun the initialization. If I try running `init` a second time I get:

```text
error: nodePort of karmada apiserver 32443 already exist
```

So it will never retry creation of the namespace it failed on. Being able to retry by running `init` to pick up where it left off would be really helpful.

**Comment by tillig** [comment]
It looks like #2492 and #2504 somewhat overlap with this concept - issues during install where it may have been pretty helpful to be able to re-run `init` and fix things up.

**Comment by RainbowMango** [comment]
cc init owners @lonelyCZ @prodanlabs for comments.
Allow `init` to fix things up sounds great. I don't know how many gaps there are in implementing this capability.

**Comment by lonelyCZ** [comment]
> Being able to retry by running init to pick up where it left off would be really helpful.

This is a good idea, but it seems to introduce inspection in each phase that could make the codes redundancy. 

> if the cluster is already initialized it'd be nice to have the init verify that things are wired up correctly and fix things that aren't.

It seems difficult to let `init` fix things that aren't. I prefer to import `pre-check` at first phase, like `kubeadm` that also firstly run `reset` and then run `init`, what do you think? @tillig

**Comment by tillig** [comment]
I don't have a super strong opinion, but my experience with tools like `istioctl` is that installation is reasonably idempotent - I can run it with the same parameters repeatedly and it will apply the installation each time. If something fails, I can run it again and it will see what's done, skip that, and apply any remaining changes. If something gets corrupted, like if you somehow delete something "important," running the install will restore it.

Similar concept with an operator - deploy a manifest with desired state, if anything gets corrupted the operator fixes it.

If the answer here is to reset things and reinstall, that's cool. If I've joined some member clusters or something (which I haven't gotten to because I'm still stuck on install), I shouldn't lose that. Since I don't know what `reset` does, I can't say... But if it's the same as `deinit` followed by `init` that's probably not great.

**Comment by tillig** [comment]
Yes, but since I started Friday and it's a long weekend in the US I haven't been able to make progress. Is there documentation on what ports are required in communication, along with source and destination? Like, "from the client running `karmadactl` to the host cluster, you need XXXXX," "from host cluster to member clusters you need YYYYY," and so on? I work in a high security environment so ports aren't just open.

**Comment by tillig** [comment]
Another interesting use case for this sort of `init` re-apply mechanism is to _change_ an existing install. If I wanted to, for example, increase the number of replicas for the API server, right now I'd _either_ have to manually hack that _or_ do a `deinit` / `init` cycle to use `karmadactl`.

As far as installation, I found that `--karmada-apiserver-advertise-address` is what I need to set, but since that's going to be contacted from my workstation to create a namespace I'd need to go to the effort of making an ingress for that, getting the firewall hole poked for the `--port` (default `32443`). I haven't gotten through the docs on locking down the Karmada API server, but I wonder if I'm headed the right direction. Perhaps I don't need the Karmada API publicly exposed at all? I'm trying to reverse-engineer the docs and various `hack` scripts, but [places like this section in the Istio + Karmada docs](https://karmada.io/docs/userguide/service/working-with-istio-on-non-flat-network#install-karmada-control-plane) link to places that don't exist anymore.

**Comment by chaosi-zju** [comment]
Hello, dear friend, are you still stuck on this issue? You seems like have two problems:

1. You want the command `karmadactl init` to be reentrant.
2. You want to solve the `unable to create Namespace` bug found in installation process.

I wanna if the problem 2 has fixed, if not, I would like to explore on it with you. (If you're also willing, please share your kubeconfig with me first~)

**Comment by tillig** [comment]
I ended up abandoning `karmadactl` entirely and installing with Helm. This got me past the namespace problem but I never was able to get members joined. I ended up switching to Azure Kubernetes Fleet Manager.

I think a lot of my challenge is the need for non standard ports. I work in a high security environment where things are very locked down. I could probably get my network and security teams to open things up between clusters but I'll never get that from my workstation (corporate network) to the cloud.

Between that, the challenge with `karmadactl` noted above, and the docs not really covering everything (so much is wrapped up in scripts but not documented!), I had to look for other solutions. It didn't seem like this was something we could operate in our environment.

**Comment by chaosi-zju** [comment]
I feel so sorry that the existing docs didn't give you a smooth experience, we will definitely reflect on your example and gradually improve our docs.

Back to your problem, you executed `karmadactl init` or `helm install` on your owner k8s cluster and failed, actually, finding the root cause is also a bit difficult for us since your environment is too special. But, from the log and description you provided , I think your environment may have no big problem, em, may be just some minor details that haven't been properly handled. And I wolud like to fully support you to achieve the installation success!

Are you willing to execute these commands below to check whether the Karmada in your environment is ok?

```shell
➜ ✗ git clone https://github.com/karmada-io/karmada
➜ ✗ cd karmada
➜ ✗ hack/local-up-karmada.sh
```

If this script can be ok, I have confidence to succeed in your real environment.

**Comment by RainbowMango** [comment]
Some remarkable info I can get from this conversion are:
1. Including `deinit` automatically within `init` can be a risky approach, just as @tillig mentioned on above, this may potentially lose the joined clusters.
2. The functionality and documentation of `karmadactl` are currently not sufficient, and unfortunately, the community now focusing on [karmada-operator](https://github.com/karmada-io/karmada/tree/master/operator). I think we still need to integrate the `karmadactl`.

Anyway, thank you very much for spotting.
